allow to make changes to nsswitch.conf

pbrezina opened the pull request:
Users can make changes to nsswitch.conf by modifying /etc/authselect/user-nsswitch.conf.

When the file is modified, changes can be applied with new command
`authselect apply-changes`.

These patches are ready to be reviewed, I just need to alter spec file to create the user
file during upgrade to new authselect version.


Resolves:
https://github.com/pbrezina/authselect/issues/51

pbrezina commented:
Read for review.

jhrozek commented on src/lib/files/system.c:
```diff
@@ -87,6 +91,209 @@ authselect_system_read_templates(const char *dirname,
+
+    reret = regcomp(&regex, RE_NSS, REG_EXTENDED | REG_NEWLINE);
+    if (reret != REG_NOERROR) {
+        printf("Unable to compile regular expression: regex error %d", reret);
```
Is it expected to use printf and not ERROR in library code?

jhrozek commented:
You might also want to look at these Coverity warnings:
```
Error: RESOURCE_LEAK (CWE-772):
authselect-0.4/src/cli/main.c:124: start: Starting defect path here.
authselect-0.4/src/cli/main.c:132: alloc_fn: Storage is returned from allocation function "authselect_profile_requirements".
authselect-0.4/src/lib/authselect_profile.c:87:5: alloc_fn: Storage is returned from allocation function "template_generate".
authselect-0.4/src/lib/util/template.c:421:9: alloc_fn: Storage is returned from allocation function "strdup".
authselect-0.4/src/lib/util/template.c:421:9: return_alloc_fn: Directly returning storage allocated by "strdup".
authselect-0.4/src/lib/authselect_profile.c:87:5: return_alloc_fn: Directly returning storage allocated by "template_generate".
authselect-0.4/src/cli/main.c:132: var_assign: Assigning: "requirements" = storage returned from "authselect_profile_requirements(profile, features)".
authselect-0.4/src/cli/main.c:165: noescape: Resource "requirements" is not freed or pointed-to in "printf".
authselect-0.4/src/cli/main.c:176: leaked_storage: Variable "requirements" going out of scope leaks the storage it points to.
#  174|       }
#  175|   
#  176|->     return ret;
#  177|   }
#  178|   

Error: RESOURCE_LEAK (CWE-772):
authselect-0.4/src/cli/main.c:377: start: Starting defect path here.
authselect-0.4/src/cli/main.c:384: alloc_fn: Storage is returned from allocation function "authselect_profile_requirements".
authselect-0.4/src/lib/authselect_profile.c:87:5: alloc_fn: Storage is returned from allocation function "template_generate".
authselect-0.4/src/lib/util/template.c:421:9: alloc_fn: Storage is returned from allocation function "strdup".
authselect-0.4/src/lib/util/template.c:421:9: return_alloc_fn: Directly returning storage allocated by "strdup".
authselect-0.4/src/lib/authselect_profile.c:87:5: return_alloc_fn: Directly returning storage allocated by "template_generate".
authselect-0.4/src/cli/main.c:384: var_assign: Assigning: "requirements" = storage returned from "authselect_profile_requirements(profile, features)".
authselect-0.4/src/cli/main.c:400: leaked_storage: Variable "requirements" going out of scope leaks the storage it points to.
#  398|       authselect_profile_free(profile);
#  399|   
#  400|->     return ret;
#  401|   }
#  402|   

Error: CLANG_WARNING:
authselect-0.4/src/lib/authselect_profile.c:199:17: warning: Value stored to 'sym_all' during its initialization is never read
#    const char *sym_all[]      = {FILES_ALL, NULL};
#                ^~~~~~~          ~~~~~~~~~~~~~~~~~
authselect-0.4/src/lib/authselect_profile.c:199:17: note: Value stored to 'sym_all' during its initialization is never read
#    const char *sym_all[]      = {FILES_ALL, NULL};
#                ^~~~~~~          ~~~~~~~~~~~~~~~~~
#  197|       const char *sym_pam[]      = {FILES_PAM, NULL};
#  198|       const char *sym_dconf[]    = {FILES_DCONF, NULL};
#  199|->     const char *sym_all[]      = {FILES_ALL, NULL};
#  200|       char **targets;
#  201|       errno_t ret;

Error: UNINIT (CWE-457):
authselect-0.4/src/lib/files/system.c:101: var_decl: Declaring variable "ret" without initializer.
authselect-0.4/src/lib/files/system.c:140: uninit_use: Using uninitialized value "ret".
#  138|   done:
#  139|       regfree(&regex);
#  140|->     if (ret != EOK) {
#  141|           string_array_free(maps);
#  142|       }

Error: COMPILER_WARNING:
authselect-0.4/src/lib/files/system.c: scope_hint: In function 'authselect_system_generate'
authselect-0.4/src/lib/files/system.c:140:8: warning: 'ret' may be used uninitialized in this function [-Wmaybe-uninitialized]
#     if (ret != EOK) {
#        ^
authselect-0.4/src/lib/files/system.c:101:13: note: 'ret' was declared here
#     errno_t ret;
#             ^~~
#  138|   done:
#  139|       regfree(&regex);
#  140|->     if (ret != EOK) {
#  141|           string_array_free(maps);
#  142|       }

Error: CLANG_WARNING:
authselect-0.4/src/lib/files/system.c:140:13: warning: The left operand of '!=' is a garbage value
#    if (ret != EOK) {
#            ^
authselect-0.4/src/lib/files/system.c:239:9: note: Assuming 'generated' is not equal to NULL
#    if (generated == NULL) {
#        ^~~~~~~~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:239:5: note: Taking false branch
#    if (generated == NULL) {
#    ^
authselect-0.4/src/lib/files/system.c:246:5: note: Control jumps to 'case 0:'  at line 247
#    switch (ret) {
#    ^
authselect-0.4/src/lib/files/system.c:248:15: note: Calling 'authselect_system_nsswitch_find_maps'
#        ret = authselect_system_nsswitch_find_maps(generated, &maps);
#              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:101:5: note: 'ret' declared without an initial value
#    errno_t ret;
#    ^~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:106:9: note: Assuming 'maps' is not equal to NULL
#    if (maps == NULL) {
#        ^~~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:106:5: note: Taking false branch
#    if (maps == NULL) {
#    ^
authselect-0.4/src/lib/files/system.c:111:9: note: Assuming 'reret' is equal to REG_NOERROR
#    if (reret != REG_NOERROR) {
#        ^~~~~~~~~~~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:111:5: note: Taking false branch
#    if (reret != REG_NOERROR) {
#    ^
authselect-0.4/src/lib/files/system.c:118:12: note: Assuming the condition is true
#    while ((reret = regexec(&regex, match_string, 2, m, 0)) == REG_NOERROR) {
#           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:118:5: note: Loop condition is true.  Entering loop body
#    while ((reret = regexec(&regex, match_string, 2, m, 0)) == REG_NOERROR) {
#    ^
authselect-0.4/src/lib/files/system.c:121:13: note: Assuming 'maps' is equal to NULL
#        if (maps == NULL) {
#            ^~~~~~~~~~~~
authselect-0.4/src/lib/files/system.c:121:9: note: Taking true branch
#        if (maps == NULL) {
#        ^
authselect-0.4/src/lib/files/system.c:122:13: note: Control jumps to line 139
#            goto done;
#            ^
authselect-0.4/src/lib/files/system.c:140:13: note: The left operand of '!=' is a garbage value
#    if (ret != EOK) {
#        ~~~ ^
#  138|   done:
#  139|       regfree(&regex);
#  140|->     if (ret != EOK) {
#  141|           string_array_free(maps);
#  142|       }

Error: CLANG_WARNING:
authselect-0.4/src/lib/files/system.c:464:30: warning: The left operand of '!=' is a garbage value
#            if (tmp_files[i] != NULL) {
#                ~~~~~~~~~~~~ ^
authselect-0.4/src/lib/files/system.c:411:9: note: Assuming 'ret' is equal to EOK
#    if (ret != EOK) {
#        ^~~~~~~~~~
authselect-0.4/src/lib/files/system.c:411:5: note: Taking false branch
#    if (ret != EOK) {
#    ^
authselect-0.4/src/lib/files/system.c:421:5: note: Loop condition is true.  Entering loop body
#    for (i = 0; generated[i].path != NULL; i++) {
#    ^
authselect-0.4/src/lib/files/system.c:425:9: note: Taking true branch
#        if (ret != EOK) {
#        ^
authselect-0.4/src/lib/files/system.c:426:13: note: Control jumps to line 462
#            goto done;
#            ^
authselect-0.4/src/lib/files/system.c:462:5: note: Taking true branch
#    if (ret != EOK) {
#    ^
authselect-0.4/src/lib/files/system.c:463:9: note: Loop condition is true.  Entering loop body
#        for (i = 0; generated[i].path != NULL; i++) {
#        ^
authselect-0.4/src/lib/files/system.c:464:30: note: The left operand of '!=' is a garbage value
#            if (tmp_files[i] != NULL) {
#                ~~~~~~~~~~~~ ^
#  462|       if (ret != EOK) {
#  463|           for (i = 0; generated[i].path != NULL; i++) {
#  464|->             if (tmp_files[i] != NULL) {
#  465|                   unlink(tmp_files[i]);
#  466|                   free(tmp_files[i]);

Error: RESOURCE_LEAK (CWE-772):
authselect-0.4/src/lib/util/textfile.c:224: alloc_fn: Storage is returned from allocation function "fopen".
authselect-0.4/src/lib/util/textfile.c:224: var_assign: Assigning: "fsource" = storage returned from "fopen(source, "r")".
authselect-0.4/src/lib/util/textfile.c:230: noescape: Resource "fsource" is not freed or pointed-to in "fileno".
authselect-0.4/src/lib/util/textfile.c:233: leaked_storage: Variable "fsource" going out of scope leaks the storage it points to.
#  231|       if (ret == -1) {
#  232|           ret = errno;
#  233|->         return ret;
#  234|       }
#  235|   
```

pbrezina commented:
> Since the last patch already bumps the version to 1.0, isn't it time to bump the library version files?
Might as well. Patch is included.

All issues should be fixed. I will work on the documentation now.

jhrozek commented:
There is a compilation warning:
```
../../../src/cli/main.c: In function ‘activate’:
../../../src/cli/main.c:171:10: warning: passing argument 1 of ‘free’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     free(requirements);
          ^~~~~~~~~~~~
In file included from ../../../src/cli/main.c:23:0:
/usr/include/stdlib.h:448:13: note: expected ‘void *’ but argument is of type ‘const char *’
 extern void free (void *__ptr) __THROW;
             ^~~~
../../../src/cli/main.c: In function ‘requirements’:
../../../src/cli/main.c:399:10: warning: passing argument 1 of ‘free’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     free(requirements);
          ^~~~~~~~~~~~
In file included from ../../../src/cli/main.c:23:0:
/usr/include/stdlib.h:448:13: note: expected ‘void *’ but argument is of type ‘const char *’
 extern void free (void *__ptr) __THROW;
             ^~~~
```

jhrozek commented:
This is probably a bug in the older PR I reviewed yesteday, but can you check if make rpms work for you? I'm getting:
```
libtool: warning: '../../src/lib/libauthselect.la' has not been installed in '/usr/lib64'
libtool: install: /usr/bin/install -p .libs/authselect /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/usr/bin/authselect
make[3]: Nothing to be done for 'install-data-am'.
make[3]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4/src/cli'
make[2]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4/src/cli'
Making install in src/tests
make[2]: Entering directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4/src/tests'
make[3]: Entering directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4/src/tests'
make[3]: Nothing to be done for 'install-exec-am'.
make[3]: Nothing to be done for 'install-data-am'.
make[3]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4/src/tests'
make[2]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4/src/tests'
make[2]: Entering directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
make[3]: Entering directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
/usr/bin/make  install-exec-hook
make[4]: Entering directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/etc/authselect
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/usr/share/authselect/vendor
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/etc/authselect/custom
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/etc/dconf/db/distro.d
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/etc/dconf/db/distro.d/locks
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/etc/pam.d
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/var/lib/authselect/backups
/usr/bin/mkdir -p /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/var/lib/authselect
make[4]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
 /usr/bin/mkdir -p '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/usr/share/doc/authselect'
 /usr/bin/install -p -m 644 COPYING README.md '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64/usr/share/doc/authselect'
make[3]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
make[2]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
make[1]: Leaving directory '/home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILD/authselect-0.4'
+ /usr/lib/rpm/find-lang.sh /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64 authselect
No translations found for authselect in /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1358.fc27.x86_64
error: Bad exit status from /var/tmp/rpm-tmp.hiA4k0 (%install)
```

jhrozek commented:
btw even after cherry-picking the patch from #74 I'm getting an error:
```
RPM build errors:
    Macro expanded in comment on line 102: %{name}.lang -f %{name}-profiles.5.lang

    Macro expanded in comment on line 147: %{name}.8.lang  -f %{name}-migration.7.lang

    File not found: /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1403.fc27.x86_64/usr/share/man/man8/authselect.8*
    File not found: /home/remote/jhrozek/devel/authselect/obj/rpmbuild/BUILDROOT/authselect-0.4-20180807.1403.fc27.x86_64/usr/share/man/man7/authselect-migration.7*
```

pbrezina commented:
I fixed the warning.

> btw even after cherry-picking the patch from #74 I'm getting an error:
You probably have some newer rpm version that produces the error. I removed  the lines completely instead of commenting them so hopefully it will work for you now.

pbrezina commented:
I found this issue in spec file: https://github.com/pbrezina/authselect/commit/6b55c58e4aba505860f537ddbbcba4b4114ee848 (pushed to master) I hope you can build rpms now.

jhrozek commented:
Yes, I was missing po4a on my workstation. I thought it was required only for translated man pages. Thanks.

pbrezina commented:
Thank you. I will open new pull request for the documentation.

Pushed to master:
* 6b55c58e4aba505860f537ddbbcba4b4114ee848
* 0cb925bfd35d83441512c03ee6dbcad70862ce17
* 66539342f20508ac0b16cd837ff4d88f2ba563e1
* ccefd9d760756cd642c9fad5fbc47d47544533fb
* f88b79039166e49627e8d4d6afa5a69fffc0c5e8
* c12348428dcf7aefb09c38b7b0b57f983bba937a
* 1c79b067804b2f7de6bb4bb1e961b116ea5a2659
* d581fdf42f2dcefb16fa6181215f601c7b203573
* b59d4f76070617a72bc094ad108f10bae11997e4
* 8ee493e9c72585a352a6fce11c7d8c9bae4dde5a
* e78c66674ee2909621c5040e9946bc133e96a885
* 5ec13b73f3ff7fa4ddae8be41259f41a0f29d217
* 050e772ebc9b53371ee46eb820cf9686191b4a17
